### PR TITLE
Implement Automod Features

### PR DESCRIPTION
**Pull request by plun1331**
## Summary

Implements some automod features

~~Based on https://github.com/discord/discord-api-docs/pull/4873 (that's all there is)~~
Noticed a new PR with more information, https://github.com/discord/discord-api-docs/pull/4860

## Checklist

- [ ] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why
- [ ] This PR fixes an issue.
- [x] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

### CODE REVIEW

**Comment by plun1331** [comment]
> Suggesting an `on_automod_alert` event that allows for custom automod alerts, logging or auto-mute if it meets a certain rule

Such an event would have to be implemented into the Discord gateway for bots to receive, so it's up to Discord to make that possible.
Discord already handles logging and blocking messages for you as well.

**Comment by plun1331** [comment]
> > Suggesting an `on_automod_alert` event that allows for custom automod alerts, logging or auto-mute if it meets a certain rule
> 
> Discord implemented a gateway event for this, it'll be called `on_auto_moderation_action_execution` in the library A little wordy, but it's consistent with the API.

Should there be a raw event for this if the guild isn't cached?

**Comment by plun1331** [comment]
About ready I think, haven't tested quite yet though, might try that this weekend.

File: discord/http.py [hunk 1]
```diff
@@ -2317,6 +2318,71 @@ def get_guild_command_permissions(
...
+        return self.request(r, json=payload)
+    
+    def edit_auto_moderation_rule(
+        self,
+        guild_id: Snowflake,
+        rule_id: Snowflake,
+        payload: automod.EditAutoModRule, 
+    ) -> Response[automod.AutoModRule]:
+        r = Route(
+            "POST",
```
Comment (baronkobama) [inline discord/http.py:2366]:
```suggestion
            "PATCH",
```

File: docs/api.rst [hunk 2]
```diff
@@ -1399,6 +1399,48 @@ to handle it, which defaults to print a traceback and ignoring the exception.
...
+
+    Called when an auto moderation action is executed.
+
+    The bot must have :attr:`~Permissions.manage_guild` to receive this, and 
+    :attr:`Intents.auto_moderation_execution` must be enabled.
+
+    :param payload: The event's data.
+    :type payload: :class:`AutoModActionExecutionEvent`
+
+
```
Comment (baronkobama) [inline docs/api.rst:1442]:
```suggestion
```

File: docs/api.rst [hunk 3]
```diff
@@ -1399,6 +1399,48 @@ to handle it, which defaults to print a traceback and ignoring the exception.
...
+
```
Comment (baronkobama) [inline docs/api.rst:1443]:
```suggestion
```

**Review by baronkobama — COMMENTED** [review]
After the first go-around with the code, this PR looks pretty solid to me. Most of the changes I've suggested are formatting-related, but a few relate to the code. Also, according to [this PR](https://github.com/discord/discord-api-docs/pull/5070), create, modify, and delete auto moderation rule supports passing reasons, so that should probably be implemented.

File: discord/automod.py [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+        If a role is not found in the guild's cache, 
+        then it will be returned as an :class:`Object`.
+        """
+        if self.guild is None:
+            return []
+        return [self.guild.get_role(role_id) or Object(role_id) for role_id in self.exempt_role_ids]
+    
+    @cached_property
+    def exempt_channels(self) -> List[Union[Union[TextChannel, ForumChannel, VoiceChannel], Object]]:
+        """List[Union[Union[TextChannel, ForumChannel, VoiceChannel], Object]]: The channels
```
[the remark this answers is not in the document]
Comment (plun1331) [reply discord/automod.py:342]: It's meant to emphasize that you can receive either one of the various channel objects, or an `Object`

File: discord/automod.py [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+    @cached_property
+    def exempt_roles(self) -> List[Union[Role, Object]]:
+        """List[Union[:class:`Role`, :class:`Object`]]: The roles that are exempt 
+        from this rule.
+        
+        If a role is not found in the guild's cache, 
+        then it will be returned as an :class:`Object`.
+        """
+        if self.guild is None:
+            return []
```
[the remark this answers is not in the document]
Comment (plun1331) [reply discord/automod.py:337]: Yes, will be added

File: discord/types/automod.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    enabled: bool
+    exempt_roles: List[Snowflake]
+    exempt_channels: List[Snowflake]
+        
+        
+class CreateAutoModRule(_CreateAutoModRuleOptional):
+    name: str
+    event_type: AutoModEventType
+    trigger_type: AutoModTriggerType
+    trigger_metadata: AutoModTriggerMetadata
```
[the remark this answers is not in the document]
Comment (plun1331) [reply discord/types/automod.py:77]: It's documented as required [here](https://github.com/discord/discord-api-docs/blob/main/docs/resources/Auto_Moderation.md#auto-moderation-rule-object)

File: discord/types/embed.py [hunk 7]
```diff
@@ -77,7 +77,7 @@ class EmbedAuthor(TypedDict, total=False):
     proxy_icon_url: str
 
 
-EmbedType = Literal["rich", "image", "video", "gifv", "article", "link"]
+EmbedType = Literal["rich", "image", "video", "gifv", "article", "link", "auto_moderation_message"]
```
[the remark this answers is not in the document]
Comment (plun1331) [reply discord/types/embed.py:80]: They opted not to document it, https://github.com/discord/discord-api-docs/pull/4860#discussion_r877075154

File: discord/automod.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (baronkobama) [reply discord/automod.py]: Oh, okay, makes sense

File: discord/types/embed.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (baronkobama) [reply discord/types/embed.py:80]: Sounds about right, all good

File: discord/types/automod.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (baronkobama) [reply discord/types/automod.py:77]: Got it, I think I was looking at the action object for some reason

File: discord/automod.py [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+    from .channel import ForumChannel, TextChannel, VoiceChannel
+    from .guild import Guild
+    from .member import Member
+    from .role import Role
+    from .state import ConnectionState
+    from .types.automod import (
+        AutoModAction as AutoModActionPayload,
+        AutoModActionMetadata as AutoModActionMetadataPayload,
+        AutoModRule as AutoModRulePayload,
+        AutoModTriggerMetadata as AutoModTriggerMetadataPayload,
```
Comment (Middledot) [inline discord/automod.py:57]:
```suggestion
```
seems to be unused

File: discord/automod.py [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+        if exempt_roles is not MISSING:
+            payload["exempt_roles"] = [r.id for r in exempt_roles]
+            
+        if exempt_channels is not MISSING:
+            payload["exempt_channels"] = [c.id for c in exempt_channels]
+            
+        if payload:
+            data = await http.edit_auto_moderation_rule(self.guild_id, self.id, payload, reason=reason)
+            return AutoModRule(state=self._state, data=data)
+
```
Comment (Middledot) [inline discord/automod.py:447]:
```suggestion
```

**Review by Middledot — COMMENTED** [review]
The enums and new AuditLogActions should be documented too

**Review by Lulalaby — DISMISSED** [review]

**Review by Lulalaby — APPROVED** [review]

**Review by Dorukyum — APPROVED** [review]

**Pull request merged**